security_options: update capabilities for oci-mode (#197)
sylabs/singularity-userdocs: security_options.rst

.. _sec:security_options:

{Singularity} can make use of various Linux kernel features to modify the
security scope and context of running containers. Non-root users may be granted
additional permissions using Linux capabilities. SELinux, AppArmor, and Seccomp
can be used to restrict the operations that can be performed by a container.

******************
Linux Capabilities
******************

Native runtime / non-OCI-Mode
=============================

In {Singularity}'s default configuration, without ``--oci``, a container started

> by started by root receives all capabilities, while a container started by a

Reviewer: ```suggestion
by root receives all capabilities, while a container started by a
```

container are specific to this user namespace. For example, ``CAP_SYS_ADMIN``
granted to an OCI-mode container does not give the user the ability to mount a
filesystem outside of the container's user namespace.

Because of this isolation of capabilities users can add and drop capabilities,
using ``--add-caps`` and ``--drop-caps``, without the need for the administrator
to have granted permission to do so with the ``singularity capabilities``
command.

OCI-mode containers do not inherit the user's own capabilities, but instead run
with a default set of capabilities that matches other OCI runtimes.

-  CAP_NET_RAW
-	CAP_NET_BIND_SERVICE
-	CAP_AUDIT_READ
-	CAP_AUDIT_WRITE
-	CAP_DAC_OVERRIDE
-	CAP_SETFCAP
-	CAP_SETPCAP
-	CAP_SETGID
-	CAP_SETUID
-	CAP_MKNOD
-	CAP_CHOWN
-	CAP_FOWNER
-	CAP_FSETID
-	CAP_KILL
-	CAP_SYS_CHROOT

When the container is entered as the root user (e.g. with ``--fakeroot``), these
default capabilities are added to the effective, permitted, and bounding sets.

When the container is entered as a non-root user, these default capabilities are
added to the bounding set.

*******************************
Security related action options
*******************************

> When starting a container with the action commands; ``shell``, ``exec``, and

Reviewer: ```suggestion
When starting a container with the action commands ``shell``, ``exec``, and
```

> ``run`` various flags allow fine grained control of security.

Reviewer: ```suggestion
``run``, various flags allow fine grained control of security.
```

``--add-caps``
==============

In the default non-OCI-mode, ``--add-caps`` will grant specified Linux
capabilities (e.g. ``CAP_NET_RAW``) to a container, provided that those
capabilities have been granted to the user by an administrator using the
``capability add`` command. This option will also accept the case insensitive
keyword ``all`` to add every capability granted by the administrator.

> In the OCI-mode, ``--add-caps`` will grant specified Linux capabilities (e.g.

Reviewer: ```suggestion
In OCI-mode, ``--add-caps`` will grant specified Linux capabilities (e.g.
```

``CAP_NET_RAW``) to the container. Because the container runs in a user
namespace, the capabilities are not effective on the host and do not have to be
granted by the administrator. The keyword ``all`` will grant all available
capabilities to the container.

``--drop-caps``
===============

In the default non-OCI-mode, the root user has a full set of capabilities when
they enter the container. You may choose to drop specific capabilities when you
initiate a container as root to enhance security.

For instance, to drop the ability for the root user to open a raw socket
inside the container:

.. code::

   $ sudo singularity exec --drop-caps CAP_NET_RAW library://centos ping -c 1 8.8.8.8
   ping: socket: Operation not permitted

In OCI-mode any user can use ``--drop-caps`` to run a container with fewer
capabilities than the default OCI capability set.

> The ``drop-caps`` option will also accept the case insensitive keyword

Reviewer: ```suggestion
The ``--drop-caps`` option will also accept the case insensitive keyword
```

binaries may allow a user to execute a command as a service account.

By default SetUID is disallowed within {Singularity} containers as a
security precaution, by mounting container filesystems as ``nosetuid.``

> In the default non-OCI-Mode, the root user can override this precaution and

Reviewer: ```suggestion
In the default non-OCI-mode, the root user can override this precaution and
```

allow SetUID binaries to behave as expected within a {Singularity} container
with the ``--allow-setuid`` option like so:

.. code::

   $ sudo singularity shell --allow-setuid some_container.sif

> In the OCI_mode, any user can permit SetUID binaries with the ``--allow-setuid``

Reviewer: ```suggestion
In OCI mode, any user can permit SetUID binaries with the ``--allow-setuid``
```

   1 packets transmitted, 1 received, 0% packet loss, time 0ms
   rtt min/avg/max/mdev = 18.838/18.838/18.838/0.000 ms

In the OCI-mode, the ``--keep-privs`` option can be used by any user. In this
mode, ``--keep-privs`` will cause the container to run inheriting the current
effective capabilities rather than using the OCI default capability set. When
entering the container as a non-root user, the capabilities are only inherited
to the bounding set.

``--no-privs``
==============

In the default non-OCI-mode, the ``--no-privs`` flag allows the root user to run
a container with all capabilities dropped, and sets the ``no_new_privs`` bit
that will prevent the container process gaining any further privilege.

> In OCI-mode, the ``--no-privs`` can be used by any user to run a container with

Reviewer: ```suggestion
In OCI-mode, the ``--no-privs`` option can be used by any user to run a container with
```

> all capabilities dropped, and sets the ``no_new_privs`` bit that will prevent

Reviewer: ```suggestion
all capabilities dropped, and to set the ``no_new_privs`` bit that will prevent
```

the container process gaining any further privilege.

``--security``
==============

> The ``--security`` flag, currently support in non-OCI-mode only, allows the

Reviewer: ```suggestion
The ``--security`` flag, currently supported in non-OCI-mode only, allows the
```

root user to leverage security modules such as SELinux, AppArmor, and seccomp

> within your {Singularity} container. You can also change the UID and GID of the

Reviewer: Nit: I found the switch from 3rd person ("... allows the root user ...") to 2nd person ("... within your {Singularity} container ...") a little awkward. If possible, it would be good to rephrase this paragraph so it's more consistent in that respect.
